paimon-core/src/main/java/org/apache/paimon/operation/AppendOnlyFileStoreRead.java

```diff
@@ -112,41 +115,65 @@ public RecordReader<InternalRow> createReader(DataSplit split) throws IOExceptio
                                 int[][] dataProjection =
                                         SchemaEvolutionUtil.createDataProjection(
                                                 tableSchema.fields(),
                                                 dataSchema.fields(),
                                                 projection);
-                                RowType rowType = dataSchema.logicalRowType();
+
                                 IndexCastMapping indexCastMapping =
                                         SchemaEvolutionUtil.createIndexCastMapping(
                                                 Projection.of(projection).toTopLevelIndexes(),
                                                 tableSchema.fields(),
                                                 Projection.of(dataProjection).toTopLevelIndexes(),
                                                 dataSchema.fields());
+
                                 List<Predicate> dataFilters =
                                         this.schemaId == key.schemaId
                                                 ? filters
                                                 : SchemaEvolutionUtil.createDataFilters(
                                                         tableSchema.fields(),
                                                         dataSchema.fields(),
                                                         filters);
+
+                                Pair<int[], RowType> partitionPair = null;
+                                if (dataSchema.partitionKeys().size() != 0) {
+                                    Pair<int[], int[][]> partitionMappping =
+                                            PartitionUtils.constructPartitionMapping(
+                                                    dataSchema, dataProjection);
+                                    dataProjection = partitionMappping.getRight();
+                                    partitionPair =
+                                            Pair.of(
+                                                    partitionMappping.getLeft(),
+                                                    dataSchema.projectedLogicalRowType(
+                                                            dataSchema.partitionKeys()));
+                                }
+
                                 return new BulkFormatMapping(
                                         indexCastMapping.getIndexMapping(),
                                         indexCastMapping.getCastMapping(),
+                                        partitionPair,
                                         formatDiscover
                                                 .discover(formatIdentifier)
                                                 .createReaderFactory(
-                                                        rowType, dataProjection, dataFilters));
+                                                        dataSchema.logicalRowType(),
```

Comment:
The type still contains Partition fields? It is weird.

Can you just modify to `FormatReaderFactory createReaderFactory(RowType projectedType, @Nullable List<Predicate> filters)`?

Reply:
Done this.